### PR TITLE
Add optional slack output to clipped gradients

### PR DESCRIPTION
**Pull request by ZsyRock**
## Summary

This PR is limited to the clipping primitive discussed in #317 and changes
only `jax_privacy/clipping.py` and `tests/gradient_clipping_test.py`.

- Add an optional positive integer `slack` argument to `clipped_fun` and
  `clipped_grad`.
- Build prefix-filled slack coordinates from the norm already computed by the
  existing clipping path, avoiding a second full-gradient norm traversal.
- Return `ClippedGradOutput(gradient, slack)` as one jointly
  sensitivity-bounded primary output. Both fields must be privatized together
  before slack is exposed.
- Preserve the original output structure and compiled path when `slack=None`.

## Design and scope

Slack construction is isolated in `_slack_from_norm`. It runs after
`pre_clipping_transform`, supports a dynamic scalar clipping bound, leaves the
gradient dtype unchanged, and emits float32 slack coordinates. The integer
`slack` argument remains static because it determines the output shape.

This first PR provides only the core private Slack Indicator primitive. It does
not add trainer integration, an adaptive threshold controller, CLI flags, or
CDF logging. A later controller can consume the privatized slack output without
changing this component.

The initial scope supports scalar global clipping. Discrete grid clipping and
PyTree clipping bounds are rejected when slack is enabled; per-leaf slack can
be added separately.

## Validation

- 2,245 tests passed across the repository's CI-compatible test shards.
- 16 module doctests passed.
- Pyink, Flake8, Pylint, Pytype, and Pyrefly passed.
- The lowered StableHLO for the disabled path is identical to upstream. The
  enabled path reuses the existing norm and adds only the slack construction
  and aggregation, not another full-gradient clipping pass.

### CODE REVIEW

**Comment by ZsyRock** [comment]
@ryan112358 This is now ready for review. It implements the single-pass clipped_grad integration discussed in #317, and all public CI checks pass.

**Comment by ryan112358** [comment]
Thanks for the work on this @ZsyRock .  This is a pretty large change and well beyond what we'd be able to review as a single PR.  Are all changes necessary?  Can some be simplified?  Is there a natural way some of the changes can be split up?

File: jax_privacy/clipping.py
Comment (ryan112358) [inline jax_privacy/clipping.py:1]: As a concrete suggestion, for this first PR I think you should focus on the changes to clipping.py and gradient_clipping_test.py only.  This will be a manageable scope to review.  In terms of code organization, it's ideal if you can create a standalone helper function that consumes the gradient norm, clipping bound, slack parameter, etc. and returns the slack vector.  Ideally, the internal changes you need to make to the existing function(s) in this file should be very minimal (check if slack param is set, if so call your helper function).  This should help with readability & maintainability.

I may have over-simplified things above, since I assume things like "rescale_to_unit_norm" and some of the other fields might require other changes, but hopefully some of the design ideas above still apply.

**Comment by ZsyRock** [comment]
Updated as requested: this PR is now a single commit changing only `jax_privacy/clipping.py` and `tests/gradient_clipping_test.py`. The implementation isolates slack construction in `_slack_from_norm` and reuses the norm returned by the existing clipping path. Trainer integration, the adaptive controller, CLI flags, and logging have been removed from this PR and can be proposed separately.

File: jax_privacy/clipping.py [hunk 1]
```diff
@@ -37,6 +37,70 @@
...
+  """Validates the clipping configurations supported by slack outputs."""
+  if slack_k is None:
+    return None
+  if isinstance(slack_k, bool) or not isinstance(slack_k, int):
+    raise TypeError('slack_k must be a Python integer or None.')
+  if slack_k <= 0:
+    raise ValueError('slack_k must be positive.')
+  if grid_scale is not None:
+    raise ValueError('slack_k is not supported with discrete grid clipping.')
+  if not jnp.isscalar(l2_clip_norm):
```
Comment (ryan112358) [inline jax_privacy/clipping.py:65]: Theoretically you could return slack on a per-leaf basis if you wanted to, probably out of scope for this PR, but maybe you can leave a TODO to flag this

File: jax_privacy/clipping.py [hunk 2]
```diff
@@ -37,6 +37,70 @@
...
+    raise ValueError('slack_k requires static scalar global clipping.')
+  try:
+    clip_norm = float(l2_clip_norm)
+  except (TypeError, ValueError) as error:
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Why do we require a static l2_clip_norm for slack to work?  I thought part of the reason to use slack was to dynamically update the l2_clip_norm across training?

File: jax_privacy/clipping.py [hunk 3]
```diff
@@ -37,6 +37,70 @@
...
+  return clip_norm
+
+
+def _slack_from_norm(
+    gradient_norm: jax.Array,
+    clipping_bound: float,
+    slack_k: int,
+    output_bound: float,
+) -> jax.Array:
+  """Constructs prefix-filled slack from an already-computed norm.
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Our formatter might complain about this docstring.  Usually it has to be a 1-line docstring or a full docstring with Args / Returns section.  I usually lean towards the 1-line docstring for private helper functions, and put the additional info in inline comments

File: jax_privacy/clipping.py [hunk 4]
```diff
@@ -424,6 +488,7 @@ def clipped_fun(
     prng_argnum: int | None = None,
     spmd_axis_name: str | None = None,
     grid_scale: int | None = None,
+    slack_k: int | None = None,
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I think I prefer the name "slack" over "slack_k" as long as the docstring is clear it has to be a int

File: jax_privacy/clipping.py [hunk 5]
```diff
@@ -550,9 +623,33 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
       value = optax.tree.cast(value, dtype)
       if grid_scale is not None:
         clipped, norm = clip_and_round_to_grid(value, l2_clip_norm, grid_scale)
+        safety_norm = norm
+      elif slack_k is not None:
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Can more of this be folded into the helper functions by any chance?

File: jax_privacy/clipping.py [hunk 6]
```diff
@@ -550,9 +623,33 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
...
+        if any(
+            not jnp.issubdtype(x.dtype, jnp.floating)
+            or jnp.finfo(x.dtype).bits < 32
+            for x in jax.tree.leaves(value)
+        ):
+          raise TypeError(
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Do you need this? I think you can return the slack in float32, but that shouldn't require the per-example grad leaves to have that dtype

File: jax_privacy/clipping.py [hunk 7]
```diff
@@ -550,9 +623,33 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
...
+              'slack_k requires float32 or wider primary outputs; use'
+              ' dtype=jnp.float32.'
+          )
+        assert slack_clip_norm is not None
+        clipped, norm = clip_pytree(
+            value, slack_clip_norm, rescale_to_unit_norm
+        )
+        output_bound = 1.0 if rescale_to_unit_norm else slack_clip_norm
+        clipped = ClippedGradOutput(
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: nit: try to fit this on two lines by writing "slack = ..." then "clipped = ClippedGradOutput(gradient, slack)"

**Review by ryan112358 — CHANGES_REQUESTED** [review]
Looking good!

File: jax_privacy/clipping.py (hunk 5 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Yes. Validation and all slack construction details are now contained in helpers. The existing clipping branch only creates the slack output after reusing its computed norm.

File: jax_privacy/clipping.py
Comment (ZsyRock) [reply jax_privacy/clipping.py:1]: Done. The PR remains limited to clipping.py and gradient_clipping_test.py. Slack construction is isolated in _slack_from_norm, and the existing clipping path only validates the option and conditionally calls the helper.

File: jax_privacy/clipping.py (hunk 2 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Removed the static-value requirement. A scalar clipping bound may now be dynamic under JIT; the test executes one compiled function with two different bounds.

File: jax_privacy/clipping.py (hunk 7 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Done. The helper result is assigned to slack_vector, then passed directly to ClippedGradOutput; the name slack is already the integer argument.

File: jax_privacy/clipping.py (hunk 6 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Removed the gradient dtype restriction. The original gradient dtype is preserved, while the slack output is float32; float16 and bfloat16 are covered by tests.

File: jax_privacy/clipping.py (hunk 1 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py:65]: Added a TODO for per-leaf slack while keeping this PR limited to scalar global clipping.

File: jax_privacy/clipping.py (hunk 3 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Changed the private helper to a one-line docstring and kept the bound argument in an inline comment.

File: jax_privacy/clipping.py (hunk 4 above)
Comment (ZsyRock) [reply jax_privacy/clipping.py]: Renamed the public argument to slack and documented that it must be a positive Python integer.

**Comment by ZsyRock** [comment]
All requested changes are addressed in af64fc1, and all public CI checks, including the heavy tests, pass. The import/copybara check is showing the generic migration error; could you please retry or inspect the internal import when convenient?

**Review by ryan112358 — APPROVED** [review]

**Comment by ZsyRock** [comment]
Thanks again for the approval. I noticed that the PR is now marked stale and is behind `main`, while the Copybara checks are green. Is it waiting on internal submission, or would you like me to rebase it onto `main`? I am happy to update it if needed.

**Review by ryan112358 — APPROVED** [review]

**Comment by ryan112358** [comment]
Thanks for the follow-up, and sorry I lost track of this.  The PR looks good to me but I want to get a second set of eyes on it given it does touch "privacy-critical" code, will try to get back to you by Wednesday if any further changes are needed.

**Comment by ZsyRock** [comment]
Hi Ryan, just following up on the internal check. I can see that `feedback/copybara` is failing, while `import/copybara` and all public checks are passing. Since the internal details are not visible externally, could you let me know whether any changes are needed on my side, or whether this is something handled internally?
